Add tutorial on Widom particle insertion

davidbbeyer opened the pull request:
Description of changes:
- Adds a tutorial that teaches users how to use the Widom particle insertion method to determine the excess chemical potential of a system. The considered system consists of a monovalent salt solution. Could potentially be used in the tutorials for the summer school, as discussed last Tuesday.

Commit:
Add tutorial on Widom particle insertion

davidbbeyer commented:
Maybe @kosovan or @pm-blanco could review it?

mebrito commented on doc/tutorials/widom_insertion/widom_insertion.ipynb:
```diff
@@ -0,0 +1,528 @@
+    "\n",
+    "MC simulations, on the other hand, do not have such limitations.\n",
+    "Even nonphysical moves can be considered, still leading to results with correct thermodynamic properties.\n",
+    "However, no valuable information of the system's dynamics can be obtained.\n",
```
```suggestion
    "However, no valuable information of the system dynamics can be obtained.\n",
```

Commit:
Add test for widom tutorial

mebrito commented on doc/tutorials/widom_insertion/widom_insertion.ipynb:
```diff
@@ -0,0 +1,528 @@
+    "\\lambda_b=2.0\\,,\\quad k_BT=1.0\\,,\\quad z_i=1.0\\,,\\quad N=N_i=50\n",
+    "\\end{align*}\n",
+    "\n",
+    "Because the investigated system is a monovalent salt the concentrations (and thus, densities) of positive and negative ions are the same.\n",
```
```suggestion
    "Since the investigated system is a monovalent salt, the concentrations (and thus, densities) of positive and negative ions are the same.\n",
```

mebrito commented on doc/tutorials/widom_insertion/widom_insertion.ipynb:
```diff
@@ -0,0 +1,528 @@
+    " \n",
+    "In the above expression, $\\Delta U$ is the energy difference of the system before and after the insertion of a test particle at a random position.\n",
+    "The value of the integral in this equation can be computed in a simulation in the canonical ensemble (e.g. using Langevin dynamics) by averaging over many configurations of the system (corresponding to the average $\\langle ...\\rangle$) and many positions of the trial insertion (corresponding to the integral).\n",
+    "\n",
```
Up to this point it isn't fully clear the idea of the method. It would be nice to clearly state it. For example: "The integral in Eq. (4) suggests a method to compute ... ". Furthermore, I would say in a few words how it is done in ESPResSo: is the integral done by  MC/Langevin dynamics combination?

mebrito commented on doc/tutorials/widom_insertion/widom_insertion.ipynb:
```diff
@@ -0,0 +1,528 @@
+    "In the above expression, $\\Delta U$ is the energy difference of the system before and after the insertion of a test particle at a random position.\n",
+    "The value of the integral in this equation can be computed in a simulation in the canonical ensemble (e.g. using Langevin dynamics) by averaging over many configurations of the system (corresponding to the average $\\langle ...\\rangle$) and many positions of the trial insertion (corresponding to the integral).\n",
+    "\n",
+    "The system under consideration here consists of $N$ monovalent ion pairs in an implicit solvent with $\\epsilon_\\text{r} = 78$, which is a simple model (``restricted primitive model'') for an aqueous salt solution of e.g. NaCl.\n",
```
From this line on, I would shift to a different section, where I would explain what we are going to simulate, what we are going to measure and against what we compare in order to test the method. The section could be titled "Simulation of aqueous salt solution" or "Excess chemical potential of a salt solution".

mebrito commented on doc/tutorials/widom_insertion/widom_insertion.ipynb:
```diff
@@ -0,0 +1,528 @@
+    "# we assume a unit system where the elementary charge, the thermal energy and the WCA length scale (sigma) are all 1\n",
+    "sigma = 3.55e-10 # Sigma in SI units\n",
+    "avo = 6.022e+23 # Avogadro's number in SI units\n",
+    "PREF = 1/(10**3 * avo * sigma**3) # Prefactor to mol/L\n",
```
`sigma`, `avo` and `PREF` could be moved down in the notebook where they are going to be used, in order to avoid confusion about the simulation units.

davidbbeyer replied:
I think this part should be left as it is for the following reason: in the exercise that follows directly below, one has to setup the system and thus set the box size and time step. This task thus already requires a notion of units.

mebrito commented on doc/tutorials/widom_insertion/widom_insertion.ipynb:
````diff
@@ -0,0 +1,528 @@
+   "metadata": {},
+   "source": [
+    "```python\n",
+    "system = espressomd.System(box_l=3 * [10.0])\n",
````
```suggestion
    "box_l_init = 10.0\n",
    "system = espressomd.System(box_l=3 * [box_l_init])\n",
```

davidbbeyer replied:
I fixed that.

mebrito commented on doc/tutorials/widom_insertion/widom_insertion.ipynb:
```diff
@@ -0,0 +1,528 @@
+    "system.time_step = 0.01\n",
+    "system.cell_system.skin = 0.4\n",
+    "\n",
+    "system.part.add(type=[types[\"Xplus\"]]*N_ION_PAIRS, pos=np.random.rand(N_ION_PAIRS, 3) * 20.0, q = [charges[\"Xplus\"]]*N_ION_PAIRS)\n",
```
```suggestion
    "system.part.add(type=[types[\"Xplus\"]]*N_ION_PAIRS, pos=np.random.rand(N_ION_PAIRS, 3) * box_l_init, q = [charges[\"Xplus\"]]*N_ION_PAIRS)\n",
```

davidbbeyer replied:
I fixed that.

mebrito commented on doc/tutorials/widom_insertion/widom_insertion.ipynb:
```diff
@@ -0,0 +1,528 @@
+    "system.cell_system.skin = 0.4\n",
+    "\n",
+    "system.part.add(type=[types[\"Xplus\"]]*N_ION_PAIRS, pos=np.random.rand(N_ION_PAIRS, 3) * 20.0, q = [charges[\"Xplus\"]]*N_ION_PAIRS)\n",
+    "system.part.add(type=[types[\"Xminus\"]]*N_ION_PAIRS, pos=np.random.rand(N_ION_PAIRS, 3) * 20.0, q = [charges[\"Xminus\"]]*N_ION_PAIRS)\n",
```
```suggestion
    "system.part.add(type=[types[\"Xminus\"]]*N_ION_PAIRS, pos=np.random.rand(N_ION_PAIRS, 3) * box_l_init, q = [charges[\"Xminus\"]]*N_ION_PAIRS)\n",
```

mebrito commented on doc/tutorials/widom_insertion/widom_insertion.ipynb:
```diff
@@ -0,0 +1,528 @@
+   "metadata": {},
+   "source": [
+    "**Exercise:**\n",
+    "* Implement a function ``warmup `` that removes potential overlaps between particles\n",
```
```suggestion
    "* Implement a function ``warmup `` that removes potential overlaps between particles and performs a warmup integration\n",
```

mebrito commented on doc/tutorials/widom_insertion/widom_insertion.ipynb:
```diff
@@ -0,0 +1,528 @@
+   "source": [
+    "**Exercise:**\n",
+    "* Implement a function ``warmup `` that removes potential overlaps between particles\n",
+    "* After the overlaps have been removed, the integrator should be switched to Velocity-Verlet (with Langevin thermostat) in order to perform a warmup integration"
```
```suggestion
    "* Notice that, after the overlaps have been removed, the integrator should be switched to Velocity-Verlet (with Langevin thermostat) in order to perform a warmup integration"
```

mebrito commented on doc/tutorials/widom_insertion/widom_insertion.ipynb:
```diff
@@ -0,0 +1,528 @@
+    "\n",
+    "\n",
+    "**Hints:**\n",
+    "* Refer to the documentation to find out how to set up Widom particle insertion ([Widom Insertion](https://https://espressomd.github.io/doc/reaction_methods.html#widom-insertion-for-homogeneous-systems))"
```
Link to documentation doesn't work

davidbbeyer replied:
I fixed it.

mebrito commented on doc/tutorials/widom_insertion/widom_insertion.ipynb:
```diff
@@ -0,0 +1,528 @@
+    "\\begin{equation}\n",
+    "\\mu_\\textrm{ex}^\\textrm{Davies}=-2\\cdot k_\\mathrm{B}T\\cdot 1.17 \\left(\\frac{\\sqrt{c_\\mathrm{{salt}}/1\\,\\mathrm{M}}}{1+\\sqrt{c_\\mathrm{{salt}}/1\\,\\mathrm{M}}}-0.2\\cdot c_\\mathrm{{salt}}/1\\,\\mathrm{M}\\right)\n",
+    "\\end{equation}\n",
+    "which results from an empirical extension of the Debye-Hückel theory.\n",
```
Add references for mu_ex^DH and mu_ex^Davies

davidbbeyer replied:
I added a reference.

mebrito reviewed (changes requested):
Here, there are some suggestions for changes. The most import would be to state clearly in the introduction what the method does and what will be done throughout the tutorial.

Commit:
Apply code review suggestions

Co-authored-by: mebrito <mariano.brito@icp.uni-stuttgart.de>

mebrito commented on doc/tutorials/widom_insertion/widom_insertion.ipynb:
```diff
@@ -0,0 +1,598 @@
+   "source": [
+    "## Widom Particle Insertion Method\n",
+    "\n",
+    "In a pure MD simulation, there is no direct way to access the chemical potential, since it is related to changes in the particle number, which is fixed in an MD simulation.\n",
```
```suggestion
    "In a pure MD simulation, there is no direct way to access the chemical potential, since it is related to changes in the particle number that is typically fixed.\n",
```

mebrito commented on doc/tutorials/widom_insertion/widom_insertion.ipynb:
```diff
@@ -0,0 +1,598 @@
+    "  \\mu_{\\alpha} = \\left(\\frac{\\partial G}{\\partial N_\\alpha} \\right)_{p,T,N_{\\beta\\neq\\alpha}} = \\left(\\frac{\\partial F}{\\partial N_\\alpha}\\right)_{V,T,N_{\\beta\\neq\\alpha}} \\quad .\\tag{1}\n",
+    "  \\label{eq:chem_pot_def}\n",
+    "\\end{equation}\n",
+    "It corresponds to the difference in the Helmholtz free energy $F$ or Gibbs free energy $G$ for bringing a particle from infinity into the system, or equally for removing one particle, thus changing the particle number $N_\\alpha$.\n",
```
```suggestion
    "and corresponds to the difference in the Helmholtz free energy $F$ or Gibbs free energy $G$ when adding or removing one particle in the system.\n",
```

mebrito commented on doc/tutorials/widom_insertion/widom_insertion.ipynb:
```diff
@@ -0,0 +1,598 @@
+    "\n",
+    "The chemical potential for species $\\alpha$ is defined as\n",
+    "\\begin{equation}\n",
+    "  \\mu_{\\alpha} = \\left(\\frac{\\partial G}{\\partial N_\\alpha} \\right)_{p,T,N_{\\beta\\neq\\alpha}} = \\left(\\frac{\\partial F}{\\partial N_\\alpha}\\right)_{V,T,N_{\\beta\\neq\\alpha}} \\quad .\\tag{1}\n",
```
```suggestion
    "  \\mu_{\\alpha} = \\left(\\frac{\\partial G}{\\partial N_\\alpha} \\right)_{p,T,N_{\\beta\\neq\\alpha}} = \\left(\\frac{\\partial F}{\\partial N_\\alpha}\\right)_{V,T,N_{\\beta\\neq\\alpha}} \\quad ,\\tag{1}\n",
```

mebrito commented on doc/tutorials/widom_insertion/widom_insertion.ipynb:
```diff
@@ -0,0 +1,598 @@
+    "  \\label{eq:mu_ex}\n",
+    "\\end{equation}\n",
+    "\n",
+    "In the above expression, $\\Delta U$ is the energy difference of the system before and after the insertion of an additional particle (\"test particle\") into the system.\n",
```
```suggestion
    "In the above expression, $\\Delta U$ is the energy difference of the system before and after the insertion of an additional particle (\"test particle\") into the system, and $\\langle...\\rangle$ is the canonical average for a system with $N$ particles.\n",
```

mebrito commented on doc/tutorials/widom_insertion/widom_insertion.ipynb:
```diff
@@ -0,0 +1,598 @@
+    "\\end{align*}\n",
+    "\n",
+    "Since the investigated system is a monovalent salt, the concentrations (and thus, densities) of positive and negative ions are the same.\n",
+    "The method of Widom insertion introduced can be used to calculate the excess chemical potential for this system.\n",
```
```suggestion
    "The introduced method of Widom insertion can be used to calculate the excess chemical potential for this system.\n",
```

mebrito commented on doc/tutorials/widom_insertion/widom_insertion.ipynb:
```diff
@@ -0,0 +1,598 @@
+   "source": [
+    "<a id='[1]'></a>[1] B. Widom. Some topics in the theory of fluids. Journal of Chemical Physics, 39:2808,\n",
+    "1963.\n",
+    "<a id='[2]'></a>[2] P. Atkins, J. de Paula and J. Keeler. Physical Chemistry. Oxford University Press, 2018."
```
```suggestion
   "<a id='[1]'></a>[1] B. Widom. Some topics in the theory of fluids. Journal of Chemical Physics, 39:2808, 1963.  \n",
    "<a id='[2]'></a>[2] P. Atkins, J. de Paula and J. Keeler. Physical Chemistry. Oxford University Press, 2018."
```

mebrito commented on doc/tutorials/widom_insertion/widom_insertion.ipynb:
```diff
@@ -0,0 +1,598 @@
+    "\n",
+    "In the above expression, $\\Delta U$ is the energy difference of the system before and after the insertion of an additional particle (\"test particle\") into the system.\n",
+    "The expression thus suggests that the excess chemical potential can be computed using the following procedure:\n",
+    "In order to sample the canonical average $\\langle...\\rangle$, one has to insert a test particle into many different configurations of the system (sampled according to the canonical distribution) and then average the associated Boltzmann factor $\\exp(-\\beta\\Delta U)$.\n",
```
```suggestion
       "The expression thus suggests that the excess chemical potential can be computed using the following procedure.\n",
    "The canonical average $\\langle\\exp(-\\beta\\Delta U)\\rangle$ can be computed by inserting a test particle into many different configurations of the system (sampled according to the canonical distribution) and then averaging the associated Boltzmann factor $\\exp(-\\beta\\Delta U)$.\n",
```

mebrito reviewed (changes requested):
Request for small wording changes in the text.

jngrad committed:
doc: Improve Widom tutorial

Provide the code for the matplotlib figure and the main simulation
loop. Detail the derivation of the analytical formula for the excess
potential (DH and Davies) and source the relevant chapter in Atkins'
textbook. Reduce the simulation time in the testsuite and make the
test deterministic by removing all but one degree of freedom in the
P3M algorithm. Flush stdout to avoid breaking stderr (tqdm progress
bars). Fix issues in LaTeX equations.

jngrad committed:
doc: Fix broken reaction methods code examples

jngrad committed:
Merge branch 'python' into python
